src/bootstrap/doc.rs

```diff
@@ -650,21 +651,22 @@ macro_rules! tool_doc {
                 let target = self.target;
                 builder.info(&format!("Documenting stage{} {} ({})", stage, stringify!($tool).to_lowercase(), target));
 
-                // This is the intended out directory for compiler documentation.
-                let out = builder.compiler_doc_out(target);
-                t!(fs::create_dir_all(&out));
-
-                let compiler = builder.compiler(stage, builder.config.build);
-
                 if !builder.config.compiler_docs {
                     builder.info("\tskipping - compiler/tool docs disabled");
                     return;
                 }
 
+                // This is the intended out directory for compiler documentation.
+                let out = builder.compiler_doc_out(target);
+                t!(fs::create_dir_all(&out));
+
                 // Build rustc docs so that we generate relative links.
                 builder.ensure(Rustc { stage, target });
+                // Rustdoc needs the rustc sysroot available to build.
+                builder.ensure(check::Rustc { target: builder.config.build });
```

Comment:
I would expect target: target here, not builder.config.build. The comment here is also a little misleading -- note that this section is shared between rustdoc and rustfmt documenting, so we should refer to something like "Rustc tools require that the rustc sysroot is present".

Reply:
Maybe I'm confused - why should this use the --target? Doesn't rustdoc need to run on the host to generate docs?

Reply:
The previous `Rustc` already does this. This one is for compiling the sysroot, which has to be for the same triple as for which you will be documenting AFAIU. In this case `target` and not `builder.config.build`.

Reply:
I would expect that `rustdoc --target foo` needs to load libraries from `$sysroot/lib/rustlib/foo/lib/...`, not `$sysroot/lib/rustlib/$build_triple/lib/...`, because it needs to be able to do things like `use crate_in_sysroot::windows_only_api;` if it's cross compiling for windows, but that's not possible if that crate isn't available in the sysroot.

I think this is the same thing @bjorn3 is saying, but not sure.

Reply:
I tried changing the target and it didn't help:
```diff
diff --git a/src/bootstrap/doc.rs b/src/bootstrap/doc.rs
index a38915ad6f0..3596bd58d01 100644
--- a/src/bootstrap/doc.rs
+++ b/src/bootstrap/doc.rs
@@ -663,7 +663,7 @@ fn run(self, builder: &Builder<'_>) {
                 // Build rustc docs so that we generate relative links.
                 builder.ensure(Rustc { stage, target });
                 // Rustdoc needs the rustc sysroot available to build.
-                builder.ensure(check::Rustc { target: builder.config.build });
+                builder.ensure(check::Rustc { target });
 
                 // Symlink compiler docs to the output directory of rustdoc documentation.
                 let compiler = builder.compiler(stage, builder.config.build);
```
```
error[E0464]: multiple matching crates for `lazy_static`
  --> src/librustdoc/lib.rs:21:1
   |
21 | extern crate lazy_static;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: candidates:
           crate `lazy_static`: /home/joshua/rustc/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblazy_static-d27507b084dee135.rlib
```

Reply:
FWIW `lazy_static` seems to be being built for both the target and host - shouldn't it only be built for the host?
```
$ fd liblazy_static-d27507b084dee135.rlib build
build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblazy_static-d27507b084dee135.rlib
build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblazy_static-d27507b084dee135.rlib
$ fd lazy_static build/x86_64-unknown-linux-gnu/stage2*
build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblazy_static-3a23310023808614.rmeta
build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblazy_static-d27507b084dee135.rlib
```

Reply:
:facepalm: I had `download-rustc` set, which doesn't support cross-compiling.